File: bratek20-architecture/src/main/java/pl/bratek20/architecture/events/impl/EventsLogic.java
package pl.bratek20.architecture.events.impl;

import com.google.common.eventbus.EventBus;
import com.google.common.eventbus.Subscribe;
import pl.bratek20.architecture.events.api.Event;
import pl.bratek20.architecture.events.api.EventListener;
import pl.bratek20.architecture.events.api.EventPublisher;

import java.lang.reflect.ParameterizedType;
import java.util.List;
import java.util.Set;

public class EventsLogic implements EventPublisher {
    private final EventBus eventBus = new EventBus();

    public EventsLogic(Set<EventListener> listeners) {
        listeners.forEach(this::subscribe);
    }

    @Override
    public void publish(Event event) {
        eventBus.post(event);
    }


    private <T extends Event> void subscribe(EventListener<T> listener) {
        // Reflection to find the actual event type handled by the listener
        ParameterizedType type = (ParameterizedType) listener.getClass().getGenericInterfaces()[0];
        Class<T> eventType = (Class<T>) type.getActualTypeArguments()[0];

        // Register a wrapper that only forwards the correct event types
        eventBus.register(new Object() {
            @Subscribe
            public void handleEvent(Object event) {
                if (eventType.isInstance(event)) {
                    listener.handleEvent(eventType.cast(event));
                }
            }
        });
    }
}

File: bratek20-architecture/src/main/kotlin/com/github/bratek20/architecture/events/impl/EventsLogic.kt
package com.github.bratek20.architecture.events.impl

import com.github.bratek20.architecture.events.api.Event
import com.github.bratek20.architecture.events.api.EventListener
import com.github.bratek20.architecture.events.api.EventPublisher
import com.google.common.eventbus.EventBus
import com.google.common.eventbus.Subscribe
import java.lang.reflect.ParameterizedType
import java.util.function.Consumer

class EventsLogic(listeners: Set<EventListener<*>>) : EventPublisher {
    private val eventBus = EventBus()

    init {
        listeners.forEach(Consumer { listener: EventListener<*> -> this.subscribe(listener) })
    }

    override fun publish(event: Event?) {
        eventBus.post(event)
    }


    private fun <T : Event?> subscribe(listener: EventListener<T>) {
        // Reflection to find the actual event type handled by the listener
        val type = listener.javaClass.genericInterfaces[0] as ParameterizedType
        val eventType = type.actualTypeArguments[0] as Class<T>

        // Register a wrapper that only forwards the correct event types
        eventBus.register(object : Any() {
            @Subscribe
            fun handleEvent(event: Any?) {
                if (eventType.isInstance(event)) {
                    listener.handleEvent(eventType.cast(event))
                }
            }
        })
    }
}
